zk/conn.go

```diff
@@ -421,23 +426,12 @@ func (c *Conn) authenticate() error {
 		return err
 	}
 
-	c.sendSetWatches()
-
-	// connect response
-
-	// package length
+	// Receive and decode a connect response.
 	c.conn.SetReadDeadline(time.Now().Add(c.recvTimeout * 10))
 	_, err = io.ReadFull(c.conn, buf[:4])
 	c.conn.SetReadDeadline(time.Time{})
 	if err != nil {
-		// Sometimes zookeeper just drops connection on invalid session data,
-		// we prefer to drop session and start from scratch when that event
-		// occurs instead of dropping into loop of connect/disconnect attempts
-		c.sessionID = 0
-		c.passwd = emptyPassword
-		c.lastZxid = 0
-		c.setState(StateExpired)
-		return ErrSessionExpired
+		return err
```

Comment:
This change was due to some weirdness in ZooKeeper where it would kill the client with no response if an invalid session ID was sent. This caused an infinite loop of reconnects.

I agree that not killing the session on an I/O error is technically correct though.. just depends on if ZooKeeper still exhibits that behavior.